packages/flutter_tools/lib/src/localizations/gen_l10n.dart

```diff
@@ -1150,17 +1147,16 @@ class LocalizationsGenerator {
           assert(node.children[1].type == ST.identifier);
           final Node identifier = node.children[1];
           // Check that placeholders exist.
-          // TODO(thkim1011): Make message.placeholders a map so that we don't need to do linear time search.
-          // See https://github.com/flutter/flutter/issues/112709
-          final Placeholder placeholder = message.placeholders.firstWhere(
-            (Placeholder placeholder) => placeholder.name == identifier.value,
-            orElse: () {
-              throw L10nException('''
-Make sure that the specified placeholder is defined in your arb file.
-$translationForMessage
-${Parser.indentForError(identifier.positionInMessage)}''');
-            }
-          );
+          final Placeholder? placeholder = message.placeholders[identifier.value];
+          if (placeholder == null) {
+            throw L10nParserException(
```

Comment:
Hopefully in a future version of gen_localizations, we'll be able to provide the line number in the input file in the L10nParserException.